File: src/sandbox/sandbox.ts
import { SandboxPolicy } from '../types/policy.js';
import { Capabilities, detectCapabilities } from '../runtime/capability-detect.js';
import { ComputeBackendSelection, selectComputeBackend } from '../runtime/backends.js';
import { assertGuestCodeSafety, normalizePolicy } from '../runtime/policy.js';
import { ProbeOptions, ProbeRegistry, SandboxTracepoint } from './probes.js';
import { KVStore, MemoryBinding } from '../memory/tiered-memory.js';
import {
  InlineExecutor,
  SandboxExecutor,
  WorkerExecutor,
  SandboxRuntimeMode,
  defaultTransportFactory,
} from './executor.js';
import { QuickJsExecutor } from '../runtime/wasm/quickjs-executor.js';
import {
  createDefaultSnapshotStore,
  SandboxSnapshot,
  SnapshotStore,
} from './snapshot-store.js';

export interface SandboxOptions {
  policy?: Partial<SandboxPolicy>;
  initialState?: Record<string, unknown>;
  store?: SnapshotStore;
  onLog?: (message: string) => void;
  /**
   * Tiered key/value memory exposed to guest code as `ctx.local` (private to
   * this sandbox) and `ctx.global` (shared across the fabric). Reading global
   * memory is I/O, so it makes replay non-deterministic — see README.
   */
  memory?: MemoryBinding;
  /**
   * Execution mode:
   * - `inline` (default): guests run in the host realm behind a token-scanned
   *   `Function` boundary; supports the live `ctx.local` / `ctx.global` tiers.
   * - `worker`: guests run over the message protocol, off the main thread when
   *   `workerFactory` is supplied; thread isolation + terminate-on-timeout.
   * - `wasm`: guests run inside QuickJS compiled to WebAssembly — a real realm
   *   boundary (no host globals reachable) with enforced memory and time
   *   limits. Requires the optional `quickjs-emscripten` package.
   *
   * `worker` and `wasm` exchange only serializable state, so they do not expose
   * the live memory tiers.
   */
  runtime?: SandboxRuntimeMode;
  /** Factory for the backing Web Worker when `runtime: 'worker'`. */
  workerFactory?: () => Worker;
}

export interface RunOptions {
  name?: string;
  args?: unknown;
  estimatedMemoryMB?: number;
}

export interface RunResult {
  id: string;
  name: string;
  ok: boolean;
  value?: unknown;
  error?: string;
  logs: string[];
  startedAt: number;
  durationMs: number;
}

export interface RunRecord {
  id: string;
  name: string;
  code: string;
  args?: unknown;
  estimatedMemoryMB: number;
  startedAt: number;
  durationMs: number;
  ok: boolean;
}

/** The only surface guest code can touch. */
export interface GuestContext {
  state: Record<string, unknown>;
  args: unknown;
  log: (message: string) => void;
  /** Sandbox-private key/value tier (present only when memory is bound). */
  local?: KVStore;
  /** Fabric-shared key/value tier (present only when memory is bound). */
  global?: KVStore;
}

function newId(): string {
  try {
    return crypto.randomUUID();
  } catch {
    return `id-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
  }
}

/**
 * A client-side sandbox: runs guest code against an isolated state object
 * under a governance policy, records every run, and supports snapshot,
 * fork, restore, and replay.
 *
 * Isolation scales with `runtime`: `inline` (token-scanned `Function`),
 * `worker` (off-thread, terminable), or `wasm` (a true QuickJS-on-WebAssembly
 * realm with no host access and enforced memory/time limits).
 */
export class Sandbox {
  readonly id: string;
  readonly policy: SandboxPolicy;
  readonly capabilities: Capabilities | null;
  readonly backend: ComputeBackendSelection;
  readonly parentSnapshotId?: string;
  readonly runtime: SandboxRuntimeMode;

  private state: Record<string, unknown>;
  private readonly initialState: Record<string, unknown>;
  private readonly store: SnapshotStore;
  private readonly executor: SandboxExecutor;
  private readonly onLog: (message: string) => void;
  private readonly runLog: RunRecord[] = [];
  private readonly probes: ProbeRegistry;
  private readonly memory?: MemoryBinding;
  private readonly options: SandboxOptions;
  private disposed = false;

  private constructor(
    options: SandboxOptions,
    capabilities: Capabilities | null,
    parentSnapshotId?: string
  ) {
    this.id = newId();
    this.options = options;
    this.policy = normalizePolicy(options.policy);
    this.capabilities = capabilities;
    this.backend = selectComputeBackend(capabilities);
    this.parentSnapshotId = parentSnapshotId;
    this.runtime = options.runtime ?? 'inline';
    this.initialState = structuredClone(options.initialState ?? {});
    this.state = structuredClone(this.initialState);
    this.store = options.store ?? createDefaultSnapshotStore();
    this.onLog = options.onLog ?? (() => {});
    this.probes = new ProbeRegistry(this.onLog);
    this.memory = options.memory;

    if (this.runtime !== 'inline' && this.memory) {
      this.onLog(`${this.runtime} runtime does not expose ctx.local/ctx.global; memory tiers are inline-only.`);
    }
    if (this.runtime === 'worker') {
      this.executor = new WorkerExecutor(this.policy, defaultTransportFactory(options.workerFactory));
    } else if (this.runtime === 'wasm') {
      this.executor = new QuickJsExecutor(this.policy);
    } else {
      this.executor = new InlineExecutor(
        this.policy,
        () => (this.memory ? { local: this.memory.local, global: this.memory.global } : {}),
        this.onLog
      );
    }
  }

  static async create(options: SandboxOptions = {}): Promise<Sandbox> {
    let capabilities: Capabilities | null = null;
    try {
      capabilities = await detectCapabilities();
    } catch {
      capabilities = null;
    }
    return new Sandbox(options, capabilities);
  }

  /** Restore a sandbox from a persisted snapshot. */
  static async restore(snapshotId: string, options: SandboxOptions = {}): Promise<Sandbox> {
    const store = options.store ?? createDefaultSnapshotStore();
    const snapshot = await store.load(snapshotId);
    if (!snapshot) {
      throw new Error(`Snapshot "${snapshotId}" not found.`);
    }
    let capabilities: Capabilities | null = null;
    try {
      capabilities = await detectCapabilities();
    } catch {
      capabilities = null;
    }
    return new Sandbox(
      { ...options, store, initialState: snapshot.state },
      capabilities,
      snapshot.id
    );
  }

  getState(): Record<string, unknown> {
    return structuredClone(this.state);
  }

  getRunLog(): RunRecord[] {
    return this.runLog.map((record) => ({ ...record }));
  }

  /**
   * Attach a verified eBPF probe program to a sandbox tracepoint. Probes at
   * `run:start` act as admission control: a nonzero return value denies the
   * run. Probes at other tracepoints observe and update their maps; return
   * values are ignored. Throws if the program fails verification.
   * See TRACEPOINT_LAYOUTS for the context struct each tracepoint exposes.
   */
  attachProbe(tracepoint: SandboxTracepoint, options: ProbeOptions): string {
    this.assertUsable();
    return this.probes.attach(tracepoint, options);
  }

  detachProbe(id: string): boolean {
    return this.probes.detach(id);
  }

  /**
   * Execute guest code. The guest sees only `ctx` (state, args, log) and a
   * frozen standard library — no network, no DOM, no ambient globals. Failures
   * are returned as `ok: false` results, never thrown.
   */
  async run(code: string, options: RunOptions = {}): Promise<RunResult> {
    this.assertUsable();
    const runId = newId();
    const runIndex = this.runLog.length;
    const name = options.name ?? `${this.policy.entrypoint}#${runIndex + 1}`;
    const estimatedMemoryMB = options.estimatedMemoryMB ?? 0;
    const logs: string[] = [];
    const startedAt = Date.now();
    const startedClock = typeof performance !== 'undefined' ? performance.now() : startedAt;

    const finish = (ok: boolean, value?: unknown, error?: string): RunResult => {
      const durationMs =
        (typeof performance !== 'undefined' ? performance.now() : Date.now()) - startedClock;
      this.runLog.push({
        id: runId,
        name,
        code,
        args: options.args,
        estimatedMemoryMB,
        startedAt,
        durationMs,
        ok,
      });
      this.probes.fire('run:end', [
        BigInt(runIndex),
        ok ? 1n : 0n,
        BigInt(Math.round(durationMs * 1000)),
        BigInt(logs.length),
        BigInt(code.length),
      ]);
      return { id: runId, name, ok, value, error, logs, startedAt, durationMs };
    };

    let workingState: Record<string, unknown>;
    try {
      assertGuestCodeSafety(code, this.policy.maxGuestCodeLength);
      workingState = structuredClone(this.state);
    } catch (error) {
      return finish(false, undefined, error instanceof Error ? error.message : String(error));
    }

    const veto = this.probes.fire('run:start', [
      BigInt(runIndex),
      BigInt(code.length),
      BigInt(Math.round(estimatedMemoryMB)),
      options.args === undefined ? 0n : 1n,
      BigInt(startedAt),
    ]);
    if (veto) {
      return finish(false, undefined, `Run denied by probe "${veto}".`);
    }

    const outcome = await this.executor.execute({
      code,
      name,
      args: options.args,
      estimatedMemoryMB,
      state: workingState,
    });

    for (const message of outcome.logs) {
      logs.push(message);
      this.onLog(message);
      this.probes.fire('log', [BigInt(runIndex), BigInt(message.length)]);
    }

    if (outcome.ok) {
      // Commit state only on success so failed runs cannot corrupt the sandbox.
      this.state = outcome.state;
      const result = finish(true, outcome.value);
      await this.maybeAutoSnapshot();
      return result;
    }
    return finish(false, undefined, outcome.error);
  }

  /** Persist current state to the snapshot store and return the snapshot. */
  async snapshot(label?: string): Promise<SandboxSnapshot> {
    this.assertUsable();
    const snapshot: SandboxSnapshot = {
      id: newId(),
      label,
      createdAt: Date.now(),
      sandboxId: this.id,
      parentSnapshotId: this.parentSnapshotId,
      runCount: this.runLog.length,
      state: structuredClone(this.state),
    };
    await this.store.save(snapshot);
    let stateBytes = 0;
    try {
      stateBytes = JSON.stringify(snapshot.state)?.length ?? 0;
    } catch {
      stateBytes = 0;
    }
    this.probes.fire('snapshot', [BigInt(this.runLog.length), BigInt(stateBytes)]);
    return snapshot;
  }

  /**
   * Branch a new sandbox from the current state. The fork gets its own copy
   * of state and an empty run log; the parent is untouched. Probes are not
   * copied — attach them to the fork explicitly if needed.
   */
  async fork(options: Omit<SandboxOptions, 'initialState'> = {}): Promise<Sandbox> {
    this.assertUsable();
    const snapshot = await this.snapshot('fork-point');
    return new Sandbox(
      {
        policy: options.policy ?? this.policy,
        store: options.store ?? this.store,
        onLog: options.onLog ?? this.onLog,
        memory: options.memory ?? this.memory,
        runtime: options.runtime ?? this.runtime,
        workerFactory: options.workerFactory ?? this.options.workerFactory,
        initialState: structuredClone(this.state),
      },
      this.capabilities,
      snapshot.id
    );
  }

  /**
   * Re-execute this sandbox's recorded runs from its initial state in a fresh
   * sandbox, returning the replayed results. Useful for debugging a run
   * sequence or verifying that behavior is reproducible.
   */
  async replay(): Promise<{ results: RunResult[]; finalState: Record<string, unknown> }> {
    this.assertUsable();
    const replayBox = new Sandbox(
      {
        policy: { ...this.policy, coldStartMs: 0, snapshotEveryNRuns: 0 },
        store: this.store,
        onLog: () => {},
        memory: this.memory,
        initialState: structuredClone(this.initialState),
      },
      this.capabilities
    );
    const results: RunResult[] = [];
    for (const record of this.runLog) {
      results.push(
        await replayBox.run(record.code, {
          name: record.name,
          args: record.args,
          estimatedMemoryMB: record.estimatedMemoryMB,
        })
      );
    }
    return { results, finalState: replayBox.getState() };
  }

  dispose(): void {
    this.disposed = true;
    this.state = {};
    this.executor.dispose();
  }

  private assertUsable(): void {
    if (this.disposed) {
      throw new Error('Sandbox has been disposed.');
    }
  }

  private async maybeAutoSnapshot(): Promise<void> {
    const cadence = this.policy.snapshotEveryNRuns;
    if (cadence > 0 && this.runLog.length % cadence === 0) {
      await this.snapshot('auto');
    }
  }
}

File: python/src/webkaya/sandbox.py
"""Client-side sandbox (Python port of ``src/sandbox/sandbox.ts``).

Runs guest Python against an isolated state dict under a governance policy,
records every run, and supports snapshot / fork / restore / replay. Guest code
runs behind a token-scanned, restricted-builtins boundary — an honesty note,
not a hard security claim. CPython cannot preempt a running thread, so the
policy ``timeout_ms`` is advisory in this local engine (enforced on the browser
and server tiers).
"""

from __future__ import annotations

import copy
import textwrap
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

from .memory import MemoryBinding
from .policy import SandboxPolicy, assert_guest_code_safety, normalize_policy

_SAFE_BUILTIN_NAMES = (
    # Data / iteration helpers
    "abs", "all", "any", "bool", "dict", "divmod", "enumerate", "filter",
    "float", "int", "len", "list", "map", "max", "min", "range", "reversed",
    "round", "set", "sorted", "str", "sum", "tuple", "zip",
    # Exception types so guests can raise and catch normally
    "Exception", "ValueError", "RuntimeError", "KeyError", "IndexError",
    "TypeError", "ZeroDivisionError", "ArithmeticError", "AssertionError",
    "StopIteration", "OverflowError",
)


def _resolve_builtin(name: str):
    source = __builtins__ if isinstance(__builtins__, dict) else vars(__builtins__)
    return source[name]


_SAFE_BUILTINS = {name: _resolve_builtin(name) for name in _SAFE_BUILTIN_NAMES}


def _new_id() -> str:
    return str(uuid.uuid4())


@dataclass
class GuestContext:
    state: Dict[str, Any]
    args: Any
    log: Callable[[str], None]
    local: Any = None   # MemoryTier when memory is bound
    shared: Any = None  # MemoryTier (global tier) when memory is bound


@dataclass
class RunResult:
    id: str
    name: str
    ok: bool
    value: Any = None
    error: Optional[str] = None
    logs: List[str] = field(default_factory=list)
    started_at: float = 0.0
    duration_ms: float = 0.0


@dataclass
class RunRecord:
    id: str
    name: str
    code: str
    args: Any
    estimated_memory_mb: float
    ok: bool


@dataclass
class Snapshot:
    id: str
    label: Optional[str]
    created_at: float
    sandbox_id: str
    parent_snapshot_id: Optional[str]
    run_count: int
    state: Dict[str, Any]


class MemorySnapshotStore:
    def __init__(self) -> None:
        self._snapshots: Dict[str, Snapshot] = {}

    def save(self, snapshot: Snapshot) -> None:
        self._snapshots[snapshot.id] = copy.deepcopy(snapshot)

    def load(self, snapshot_id: str) -> Optional[Snapshot]:
        found = self._snapshots.get(snapshot_id)
        return copy.deepcopy(found) if found else None

    def list(self) -> List[Snapshot]:
        return sorted((copy.deepcopy(s) for s in self._snapshots.values()), key=lambda s: s.created_at)

    def remove(self, snapshot_id: str) -> None:
        self._snapshots.pop(snapshot_id, None)


def _compile_guest(code: str):
    body = textwrap.indent(code if code.strip() else "pass", "    ")
    source = "def _guest(ctx):\n" + body
    namespace: Dict[str, Any] = {}
    exec(compile(source, "<guest>", "exec"), {"__builtins__": _SAFE_BUILTINS}, namespace)  # noqa: S102
    return namespace["_guest"]


class Sandbox:
    def __init__(
        self,
        policy: Optional[dict] = None,
        initial_state: Optional[Dict[str, Any]] = None,
        store: Optional[MemorySnapshotStore] = None,
        on_log: Optional[Callable[[str], None]] = None,
        memory: Optional[MemoryBinding] = None,
        _parent_snapshot_id: Optional[str] = None,
    ) -> None:
        self.id = _new_id()
        self.policy: SandboxPolicy = normalize_policy(policy)
        self.parent_snapshot_id = _parent_snapshot_id
        self._initial_state = copy.deepcopy(initial_state or {})
        self._state = copy.deepcopy(self._initial_state)
        self._store = store or MemorySnapshotStore()
        self._on_log = on_log or (lambda _msg: None)
        self._memory = memory
        self._run_log: List[RunRecord] = []
        self._disposed = False

    # --- Parity aliases with the TS async API (create/restore are factories) ---
    @classmethod
    def create(cls, **kwargs) -> "Sandbox":
        return cls(**kwargs)

    @classmethod
    def restore(cls, snapshot_id: str, store: MemorySnapshotStore, **kwargs) -> "Sandbox":
        snapshot = store.load(snapshot_id)
        if snapshot is None:
            raise ValueError(f'Snapshot "{snapshot_id}" not found.')
        return cls(initial_state=snapshot.state, store=store, _parent_snapshot_id=snapshot.id, **kwargs)

    def get_state(self) -> Dict[str, Any]:
        return copy.deepcopy(self._state)

    def get_run_log(self) -> List[RunRecord]:
        return list(self._run_log)

    def run(self, code: str, name: Optional[str] = None, args: Any = None,
            estimated_memory_mb: float = 0) -> RunResult:
        self._assert_usable()
        run_id = _new_id()
        run_index = len(self._run_log)
        name = name or f"{self.policy.entrypoint}#{run_index + 1}"
        logs: List[str] = []
        started_at = time.time()
        started_clock = time.perf_counter()

        def finish(ok: bool, value: Any = None, error: Optional[str] = None) -> RunResult:
            duration_ms = (time.perf_counter() - started_clock) * 1000
            self._run_log.append(RunRecord(run_id, name, code, args, estimated_memory_mb, ok))
            return RunResult(run_id, name, ok, value, error, logs, started_at, duration_ms)

        try:
            assert_guest_code_safety(code, self.policy.max_guest_code_length)
            if estimated_memory_mb > self.policy.memory_budget_mb:
                raise ValueError(
                    f"Invocation requires ~{estimated_memory_mb}MB, above sandbox budget "
                    f"{self.policy.memory_budget_mb}MB."
                )
            working_state = copy.deepcopy(self._state)
            guest = _compile_guest(code)
        except Exception as error:  # noqa: BLE001
            return finish(False, error=str(error))

        def log(message) -> None:
            logs.append(str(message))
            self._on_log(str(message))

        ctx = GuestContext(
            state=working_state,
            args=args,
            log=log,
            local=self._memory.local if self._memory else None,
            shared=self._memory.shared if self._memory else None,
        )

        retries = max(0, self.policy.retry_count)
        last_error: Optional[str] = None
        for _attempt in range(retries + 1):
            try:
                value = guest(ctx)
                self._state = working_state  # commit only on success
                result = finish(True, value=value)
                self._maybe_auto_snapshot()
                return result
            except Exception as error:  # noqa: BLE001
                last_error = str(error)
                working_state = copy.deepcopy(self._state)
                ctx.state = working_state
        return finish(False, error=last_error)

    def snapshot(self, label: Optional[str] = None) -> Snapshot:
        self._assert_usable()
        snapshot = Snapshot(
            id=_new_id(),
            label=label,
            created_at=time.time(),
            sandbox_id=self.id,
            parent_snapshot_id=self.parent_snapshot_id,
            run_count=len(self._run_log),
            state=copy.deepcopy(self._state),
        )
        self._store.save(snapshot)
        return snapshot

    def fork(self, **overrides) -> "Sandbox":
        self._assert_usable()
        snapshot = self.snapshot("fork-point")
        return Sandbox(
            policy=overrides.get("policy") or self.policy.__dict__,
            store=overrides.get("store") or self._store,
            on_log=overrides.get("on_log") or self._on_log,
            memory=overrides.get("memory") or self._memory,
            initial_state=copy.deepcopy(self._state),
            _parent_snapshot_id=snapshot.id,
        )

    def replay(self):
        self._assert_usable()
        replay_policy = dict(self.policy.__dict__)
        replay_policy.update(cold_start_ms=0, snapshot_every_n_runs=0)
        replay_box = Sandbox(
            policy=replay_policy,
            store=self._store,
            memory=self._memory,
            initial_state=copy.deepcopy(self._initial_state),
        )
        results = [
            replay_box.run(r.code, name=r.name, args=r.args, estimated_memory_mb=r.estimated_memory_mb)
            for r in self._run_log
        ]
        return results, replay_box.get_state()

    def dispose(self) -> None:
        self._disposed = True
        self._state = {}

    def _assert_usable(self) -> None:
        if self._disposed:
            raise RuntimeError("Sandbox has been disposed.")

    def _maybe_auto_snapshot(self) -> None:
        cadence = self.policy.snapshot_every_n_runs
        if cadence > 0 and len(self._run_log) % cadence == 0:
            self.snapshot("auto")
